common/djangoapps/third_party_auth/samlproviderdata/views.py

```diff
@@ -0,0 +1,54 @@
+"""
+    Viewset for auth/saml/v0/samlproviderdata
+"""
+
+from edx_rbac.mixins import PermissionRequiredMixin
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+from enterprise.models import EnterpriseCustomerIdentityProvider
+from rest_framework import permissions, viewsets
+from rest_framework.authentication import SessionAuthentication
+
+from openedx.features.enterprise_support.utils import fetch_enterprise_customer_by_id
+
+from ..models import SAMLProviderConfig, SAMLProviderData
+from .serializers import SAMLProviderDataSerializer
+
+
+class SAMLProviderDataMixin(object):
+    authentication_classes = [JwtAuthentication, SessionAuthentication]
+    permission_classes = [permissions.IsAuthenticated]
+    serializer_class = SAMLProviderDataSerializer
+
+
+class SAMLProviderDataViewSet(PermissionRequiredMixin, SAMLProviderDataMixin, viewsets.ModelViewSet):
+    """
+    A View to handle SAMLProviderData CRUD.
+    Uses the edx-rbac mixin PermissionRequiredMixin to apply enterprise authorization
+
+    Usage:
+        [HttpVerb] /auth/saml/v0/providerdata/
+    """
+    permission_required = 'enterprise.can_access_admin_dashboard'
+
+    def get_queryset(self):
+        """
+        Find and return the matching providerid for the given enterprise uuid
+        Note: There is no direct association between samlproviderdata and enterprisecustomer.
+        So we make that association in code via samlproviderdata > samlproviderconfig ( via entity_id )
+        then, we fetch enterprisecustomer via samlproviderconfig > enterprisecustomer ( via association table )
+        """
+        enterprise_customer_idp = EnterpriseCustomerIdentityProvider.objects.get(
+            enterprise_customer__uuid=self.requested_enterprise_uuid
+        )
+        saml_provider = SAMLProviderConfig.objects.get(pk=enterprise_customer_idp.provider_id)
+        return SAMLProviderData.objects.filter(entity_id=saml_provider.entity_id)
+
+    @property
+    def requested_enterprise_uuid(self):
+        return self.request.query_params.get('enterprise_customer_uuid')
```

Comment:
Same as above, I think this is a very strange access pattern where we're required to specify a query string param on every request even when we're not querying for data. 

As above, I'd like to at least see some error handling around if this qs isn't provided if we decide to keep this.

Reply:
yeah for this, I agree this design is awkward.. should we change to 


* create new config with: `POST /auth/saml/v0/providerconfig/enterprise_customer_id/<uuid>/` postData
* get existing config with `GET /auth/saml/v0/providerconfig/enterpriser_customer_id/<uuid>/`
* patch/update existing with: `PATCH /auth/saml/v0/providerconfig/enterpriser_customer_id/<uuid>/` postData

then the url param really won't be needed at all ... 

thoughts?

Reply:
error handling: totally agreed. To add, I think there is room for url endpoints that look like (in the future) `/auth/saml/v0/providerconfig/<pk>` as well... these endpoints will lookup providerconfig by enterprise customer uuid so we need some thing to express that in the url. We were using query string uuid as a way to do that since the uuid here is a look param, without which we currently get a 500 (but @taliaedX is going to address that as part of a upcoming PR, we tasked it out), so I think it may could go with the query string for now, but I would rather follow conventions in api to be consistent as much as possible, if the above is a better design...


url design: on this point, totally should reconsider design, but given today is the last day of the sprint, shall we go with this design for now after fixing up the test issues, and do this rework in another PR? I think it is important right now to fix those test issues.. but agree that we may want to revisit the url, I can post in the team channel for discussion

Reply:
> yeah for this, I agree this design is awkward.. should we change to
> 
> * create new config with: `POST /auth/saml/v0/providerconfig/enterprise_customer_id/<uuid>/` postData
> * get existing config with `GET /auth/saml/v0/providerconfig/enterpriser_customer_id/<uuid>/`
> * patch/update existing with: `PATCH /auth/saml/v0/providerconfig/enterpriser_customer_id/<uuid>/` postData
> 
> then the url param really won't be needed at all ...
> 
> thoughts?

POST /auth/saml/v0/provider_config/ {postData}
GET /auth/saml/v0/provider_config/ -> Returns all provider_configs that the user has access to
GET /auth/saml/v0/provider_config/pk -> Returns singular provider_config specified by primary key
GET /auth/saml/v0/provider_config?enterprise_customer_id= -> Returns all provider_configs that match enterprise_customer_id
PATCH /auth/saml/v0/provider_config/pk {patchData}

Reply:
Another thought: I could make these changes as part of a v1/ api, it would be a good way to exercise the versioning too... just wanted to ensure I can fix up the issues, as well as keep the frontend work going as well... I can delete the v0 endpoints once we adapt v1 I guess?

Reply:
> error handling: totally agreed. To add, I think there is room for url endpoints that look like (in the future) `/auth/saml/v0/providerconfig/<pk>` as well... these endpoints will lookup providerconfig by enterprise customer uuid so we need some thing to express that in the url. We were using query string uuid as a way to do that since the uuid here is a look param, without which we currently get a 500 (but @taliaedX is going to address that as part of a upcoming PR, we tasked it out), so I think it may could go with the query string for now, but I would rather follow conventions in api to be consistent as much as possible, if the above is a better design...
> 
> url design: on this point, totally should reconsider design, but given today is the last day of the sprint, shall we go with this design for now after fixing up the test issues, and do this rework in another PR? I think it is important right now to fix those test issues.. but agree that we may want to revisit the url, I can post in the team channel for discussion

Personally I don't care about our sprint bounds, I think it's important to discuss what our "definition of done" is. If we're worried about claiming these points for the sprint, I'd argue that this ticket isn't complete if there's follow up work, so we wouldn't get those points anyways. Let's just land this how we expect it to operate in production once we're feature complete, I think there's more value in that than pushing out something we're not committed to.

Reply:
> Another thought: I could make these changes as part of a v1/ api, it would be a good way to exercise the versioning too... just wanted to ensure I can fix up the issues, as well as keep the frontend work going as well... I can delete the v0 endpoints once we adapt v1 I guess?

We could, but I think we could argue the case then that v0 doesn't provide a lot of value for us if we're landing this with the mentality that this will already be re-written (and I assume soon with the frontend work needing done). 

For frontend work I think we can stub/mock a lot if not all of the data contracts now that we have the models solidified. We know what required fields we have are, we know what data we'll be retrieving and the form it should have, so I don't think any of that work is blocked on landing the API. Definitely something we should discuss so in case things do diverge it isn't a massive overhaul, but things like changing URLs to write to/read from is easy, especially if the returned dataset doesn't change.

Reply:
discussed, will submit new PR with updates to not require query param for post